File: typescript/test/regex_pathological.test.ts
// VERSION: @voxgig/struct 0.1.0
//
// Discovery test: pathological regex inputs run against the port's re_* API.
// Each case wraps the call so one failure does not mask the others.
// The panel is the same in every port (see REGEX.md).

import { test } from 'node:test'

import {
  re_compile, re_test, re_find, re_find_all, re_replace,
} from '../dist/StructUtility'

function rep(s: string, n: number): string {
  return new Array(n + 1).join(s)
}

function record(label: string, fn: () => unknown): void {
  const t0 = process.hrtime.bigint()
  let outcome: string
  try {
    const r = fn()
    outcome = `OK | ${JSON.stringify(r)}`
  } catch (e: any) {
    outcome = `ERR | ${e && e.message ? e.message : String(e)}`
  }
  const ms = Number(process.hrtime.bigint() - t0) / 1e6
  // eslint-disable-next-line no-console
  console.log(`[regex-discovery] ${label} | ${ms.toFixed(2)}ms | ${outcome}`)
}

test('regex pathological discovery', () => {
  const A22 = rep('a', 22)
  const NEST40 = rep('(', 40) + 'a' + rep(')', 40)

  record('P1_redos_nested_plus',     () => re_test('^(a+)+$', A22 + '!'))
  record('P2_redos_alt_overlap',     () => re_test('^(a|aa)+$', A22 + '!'))
  record('P3_empty_repeat_replace',  () => re_replace('a*', 'abc', 'X'))
  record('P4_unicode_replace_dot',   () => re_replace('\\.', 'café.au.lait', '/'))
  record('P5_unicode_find_codepoint',() => re_find('é', 'café au lait'))
  record('P6_deep_nesting_compile',  () => re_test(NEST40, 'a'))
  record('P7_big_bounded_quantifier',() => re_test('^a{0,10000}b$', rep('a', 10) + 'b'))
  record('P8_invalid_pattern',       () => re_compile('[abc'))
  record('P9_backref_re2_forbidden', () => re_test('^(a+)\\1$', 'aaaa'))
  record('P10_find_all_zero_width',  () => re_find_all('a*', 'bbb'))
})

File: java/src/test/RegexPathologicalTest.java
// RUN: mvn -Dtest=RegexPathologicalTest test
//
// Discovery test: pathological regex inputs run against the port's re* API.
// Goal is to surface failures across ports, not to assert behaviour.
// Panel is the same in every port (see REGEX.md).

package voxgig.struct;

import com.google.gson.Gson;
import org.junit.jupiter.api.Test;

import java.util.function.Supplier;

class RegexPathologicalTest {
  private static final Gson GSON = new Gson();

  private static void record(String label, Supplier<Object> fn) {
    long t0 = System.nanoTime();
    String outcome;
    try {
      Object r = fn.get();
      outcome = "OK | " + GSON.toJson(r);
    } catch (Throwable e) {
      outcome = "ERR | " + e.getClass().getSimpleName() + ": " + e.getMessage();
    }
    double ms = (System.nanoTime() - t0) / 1e6;
    System.out.printf("[regex-discovery] %s | %.2fms | %s%n", label, ms, outcome);
  }

  @Test
  void panel() {
    String a22 = "a".repeat(22);
    String nest40 = "(".repeat(40) + "a" + ")".repeat(40);

    record("P1_redos_nested_plus",      () -> Struct.reTest("^(a+)+$", a22 + "!"));
    record("P2_redos_alt_overlap",      () -> Struct.reTest("^(a|aa)+$", a22 + "!"));
    record("P3_empty_repeat_replace",   () -> Struct.reReplace("a*", "abc", "X"));
    record("P4_unicode_replace_dot",    () -> Struct.reReplace("\\.", "café.au.lait", "/"));
    record("P5_unicode_find_codepoint", () -> Struct.reFind("é", "café au lait"));
    record("P6_deep_nesting_compile",   () -> Struct.reTest(nest40, "a"));
    record("P7_big_bounded_quantifier", () -> Struct.reTest("^a{0,10000}b$", "a".repeat(10) + "b"));
    record("P8_invalid_pattern",        () -> Struct.reCompile("[abc"));
    record("P9_backref_re2_forbidden",  () -> Struct.reTest("^(a+)\\1$", "aaaa"));
    record("P10_find_all_zero_width",   () -> Struct.reFindAll("a*", "bbb"));
  }
}
